### ISSUE TITLE
Compatibility with higher gcc versions/standards

### ISSUE THREAD

**Issue by mikrosk** [freemint/mintlib#18]
I have been experimenting with gcc 7.1 and noticed that sometime around gcc 5.0 the default standard has been changed to `gnu11` (instead of previous `gnu89`). This leads to the following errors due to the [crazy inline semantics changes](https://stackoverflow.com/a/216546/21009):
```
m68k-atari-mint-gcc -Wall -O2 -fomit-frame-pointer -nostdinc -I../stdio -I.. -I../include -I../mintlib -I../stdlib -I/home/mikro/gnu-tools/m68000/lib/gcc/m68k-atari-mint/7.1.0/include -I/home/mikro/gnu-tools/m68000/lib/gcc/m68k-atari-mint/7.1.0/include-fixed -I/usr/m68k-atari-mint/include -DHAVE_CONFIG_H -D_LIBC -D_REENTRANT -Wp,-MD,../.deps/linewrap.P -c ../stdio/linewrap.c -o linewrap.o
../stdio/linewrap.c:379:1: error: redefinition of 'line_wrapped'
 line_wrapped (FILE *stream)
 ^~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:111:1: note: previous definition of 'line_wrapped' was here
 line_wrapped (FILE *stream)
 ^~~~~~~~~~~~
../stdio/linewrap.c:429:1: error: redefinition of 'line_wrap_lmargin'
 line_wrap_lmargin (FILE *stream)
 ^~~~~~~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:118:1: note: previous definition of 'line_wrap_lmargin' was here
 line_wrap_lmargin (FILE *stream)
 ^~~~~~~~~~~~~~~~~
../stdio/linewrap.c:439:1: error: redefinition of 'line_wrap_set_lmargin'
 line_wrap_set_lmargin (FILE *stream, size_t lmargin)
 ^~~~~~~~~~~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:128:1: note: previous definition of 'line_wrap_set_lmargin' was here
 line_wrap_set_lmargin (FILE *stream, size_t lmargin)
 ^~~~~~~~~~~~~~~~~~~~~
../stdio/linewrap.c:454:1: error: redefinition of 'line_wrap_rmargin'
 line_wrap_rmargin (FILE *stream)
 ^~~~~~~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:143:1: note: previous definition of 'line_wrap_rmargin' was here
 line_wrap_rmargin (FILE *stream)
 ^~~~~~~~~~~~~~~~~
../stdio/linewrap.c:464:1: error: redefinition of 'line_wrap_set_rmargin'
 line_wrap_set_rmargin (FILE *stream, size_t rmargin)
 ^~~~~~~~~~~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:153:1: note: previous definition of 'line_wrap_set_rmargin' was here
 line_wrap_set_rmargin (FILE *stream, size_t rmargin)
 ^~~~~~~~~~~~~~~~~~~~~
../stdio/linewrap.c:479:1: error: redefinition of 'line_wrap_wmargin'
 line_wrap_wmargin (FILE *stream)
 ^~~~~~~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:168:1: note: previous definition of 'line_wrap_wmargin' was here
 line_wrap_wmargin (FILE *stream)
 ^~~~~~~~~~~~~~~~~
../stdio/linewrap.c:489:1: error: redefinition of 'line_wrap_set_wmargin'
 line_wrap_set_wmargin (FILE *stream, size_t wmargin)
 ^~~~~~~~~~~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:178:1: note: previous definition of 'line_wrap_set_wmargin' was here
 line_wrap_set_wmargin (FILE *stream, size_t wmargin)
 ^~~~~~~~~~~~~~~~~~~~~
../stdio/linewrap.c:505:1: error: redefinition of 'line_wrap_point'
 line_wrap_point (FILE *stream)
 ^~~~~~~~~~~~~~~
In file included from ../stdio/linewrap.c:25:0:
../stdio/linewrap.h:194:1: note: previous definition of 'line_wrap_point' was here
 line_wrap_point (FILE *stream)
 ^~~~~~~~~~~~~~~
make[1]: *** [../buildrules:267: linewrap.o] Error 1
```
Currently I can fix it by adding `-std=gnu89` into `configvars` but generally speaking, any proposal how to handle cases like this? `#ifdef` for both standards or as soon as native gcc > 4.6.4 is built ("maybe" by myself ;)) we just switch, fix and move on?

**Comment by mikrosk** [comment]
More information here: https://gcc.gnu.org/gcc-5/porting_to.html and here: https://en.wikipedia.org/wiki/Inline_function#gnu89

**Comment by th-otto** [comment]
@mikrosk 

> any proposal how to handle cases like this? #ifdef for both standards or as soon as native gcc > 4.6.4 is built ("maybe" by myself ;)) we just switch, fix and move on?

"fixing" it so that it only works with newer versions of gcc is imho a bad idea. That would force everyone to install that newer version for compiling mintlib, which in turn might give similar problems when compiling other packages.

Generally, i think there are a few solutions to that inline mess:

- do something like
#if defined __cplusplus
# define __extern_inline inline
#elif defined __GNUC_STDC_INLINE__
# define __extern_inline extern __inline __attribute__ ((__gnu_inline__))
#else
# define __extern_inline extern __inline
#endif
then make sure everywhere that define is used.

- always use static inline where it is really important for speed reasons, and don't use inline at all otherwise.

- do some complicated defines per header where extern inline is used, define it to just "inline" for most cases, and make sure that a single source file defines it to "extern inline". You may look at the gettext source (or some other package using the gnulib modules) to see how it is handled there.

I would prefer the 1st solution, this seems to be the simplest, requiring almost no changes.

PS.: i have done that support for gcc7 already some months ago, you can find it in the [gcc-7-mint branch](https://github.com/th-otto/m68k-atari-mint-gcc/commits/gcc-7-mint) of my fork. If you want, i can assemble some archive with a patch and a build script similar to Vincent's one for gcc-4.6.4.

**Comment by mikrosk** [comment]
@th-otto if you look at the mentioned functions, the biggest problem is that they use one specific feature of the gnu89 inline -- multiple body definitions. This isn't allowed in C99 at all. So supporting both standards isn't just a matter of a simple `#if` / `#else` :-(

Maybe you're right, maybe we should perhaps rewrite the problematic functions as `static inline` and the problem would disappear.

Btw I'm well aware of your work, where do you think I have the patch from. ;-) I wanted to provide a PR to @vinriviere as soon as I'm done with testing (and patching mintlib & friends -- some minor changes are needed). It's really a shame that the community isn't aware of your great work and celebrating this [unholy mess](https://bitbucket.org/ggnkua/bigbrownbuild/src/6445467e1282dd3d3f9bd238ec6c5d2787c46f88/bigbrownbuild.sh?at=default&fileviewer=file-view-default) instead. I'd like to change that and provide a standard native m68k-atari-mint-gcc 7.1.0.

**Comment by atic-atac** [comment]
Are we upstreaming our GCC 7.x patches, I'd like to see our stuff mainlined.

**Comment by th-otto** [comment]
@mikrosk 

> if you look at the mentioned functions, the biggest problem is that they use one specific feature of the gnu89 inline -- multiple body definitions

With solution one that won't be neccessary. Defining the body twice is a bad idea anyway, even if they are identical.

@atic-atac 

> Are we upstreaming our GCC 7.x patches, I'd like to see our stuff mainlined.

That would be nice, but i'm not sure wether the gcc maintainers are willing to accept patches for such an ancient platform. And it would only make sense if we get the patches to binutils upstreamed, too. Remember that there have been thoughts of removing a.out support completely.

**Comment by mikrosk** [comment]
Fixed by #19 for now.
